### ISSUE TITLE
functest-run-suite is not closing its event loop properly

### ISSUE THREAD

**Issue by thedac** [openstack-charmers/zaza#135]
After a successfull run of functest-run-suite zaza spews the bellow tracebacks. It may not be closing the event loop properly or it is trying to close it to often.



ok                                                                                                                                                                      
test_pause_resume (zaza.charm_tests.keystone.tests.CharmOperationTest)                                                                                                  
Run pause and resume tests. ... ok                                                                                                                                      
                                                                                                                                                                        
----------------------------------------------------------------------
Ran 2 tests in 149.459s

OK
ERROR:asyncio:Fatal write error on socket transport
protocol: <asyncio.sslproto.SSLProtocol object at 0x7efe44bf9080>
transport: <_SelectorSocketTransport fd=9>
Traceback (most recent call last):
  File "/usr/lib/python3.6/asyncio/selector_events.py", line 765, in write
    n = self._sock.send(data)
OSError: [Errno 9] Bad file descriptor
ERROR:asyncio:Fatal error on SSL transport
protocol: <asyncio.sslproto.SSLProtocol object at 0x7efe44bf9080>
transport: <_SelectorSocketTransport closing fd=9>
Traceback (most recent call last):
  File "/usr/lib/python3.6/asyncio/selector_events.py", line 765, in write
    n = self._sock.send(data)
OSError: [Errno 9] Bad file descriptor

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/lib/python3.6/asyncio/sslproto.py", line 650, in _process_write_backlog
    self._transport.write(chunk)
  File "/usr/lib/python3.6/asyncio/selector_events.py", line 769, in write
    self._fatal_error(exc, 'Fatal write error on socket transport')
  File "/usr/lib/python3.6/asyncio/selector_events.py", line 645, in _fatal_error
    self._force_close(exc)
  File "/usr/lib/python3.6/asyncio/selector_events.py", line 657, in _force_close
    self._loop.call_soon(self._call_connection_lost, exc)
  File "/usr/lib/python3.6/asyncio/base_events.py", line 575, in call_soon
    self._check_closed()
  File "/usr/lib/python3.6/asyncio/base_events.py", line 358, in _check_closed
    raise RuntimeError('Event loop is closed')
RuntimeError: Event loop is closed
ERROR:asyncio:Task was destroyed but it is pending!
task: <Task pending coro=<WebSocketCommonProtocol.transfer_data() running at /home/thedac/dev/run/.tox/clients/lib/python3.6/site-packages/websockets/protocol.py:528> w
ait_for=<Future pending cb=[<TaskWakeupMethWrapper object at 0x7efe3ebb5948>()]> cb=[<TaskWakeupMethWrapper object at 0x7efe44c01048>(), _wait.<locals>._on_completion()
 at /usr/lib/python3.6/asyncio/tasks.py:380]>
ERROR:asyncio:Task was destroyed but it is pending!
task: <Task pending coro=<WebSocketCommonProtocol.close_connection() running at /home/thedac/dev/run/.tox/clients/lib/python3.6/site-packages/websockets/protocol.py:805
> wait_for=<Task pending coro=<WebSocketCommonProtocol.transfer_data() running at /home/thedac/dev/run/.tox/clients/lib/python3.6/site-packages/websockets/protocol.py:5
28> wait_for=<Future pending cb=[<TaskWakeupMethWrapper object at 0x7efe3ebb5948>()]> cb=[<TaskWakeupMethWrapper object at 0x7efe44c01048>(), _wait.<locals>._on_complet
ion() at /usr/lib/python3.6/asyncio/tasks.py:380]>>
Exception ignored in: <generator object WebSocketCommonProtocol.close_connection at 0x7efe44bfb360>
Traceback (most recent call last):
  File "/home/thedac/dev/run/.tox/clients/lib/python3.6/site-packages/websockets/protocol.py", line 855, in close_connection
    if (yield from self.wait_for_connection_lost()):
  File "/home/thedac/dev/run/.tox/clients/lib/python3.6/site-packages/websockets/protocol.py", line 880, in wait_for_connection_lost
    self.timeout, loop=self.loop)
  File "/usr/lib/python3.6/asyncio/tasks.py", line 342, in wait_for
    timeout_handle = loop.call_later(timeout, _release_waiter, waiter)
  File "/usr/lib/python3.6/asyncio/base_events.py", line 544, in call_later
    timer = self.call_at(self.time() + delay, callback, *args)
  File "/usr/lib/python3.6/asyncio/base_events.py", line 554, in call_at
    self._check_closed()
  File "/usr/lib/python3.6/asyncio/base_events.py", line 358, in _check_closed
    raise RuntimeError('Event loop is closed')
RuntimeError: Event loop is closed
ERROR:asyncio:Task was destroyed but it is pending!
task: <Task pending coro=<_Task.start.<locals>.run() running at /home/thedac/dev/run/.tox/clients/lib/python3.6/site-packages/juju/client/connection.py:565> wait_for=<F
uture pending cb=[<TaskWakeupMethWrapper object at 0x7efe3fa48a98>()]>>
ERROR:asyncio:Task was destroyed but it is pending!
task: <Task pending coro=<_Task.start.<locals>.run() running at /home/thedac/dev/run/.tox/clients/lib/python3.6/site-packages/juju/client/connection.py:565> wait_for=<F
uture pending cb=[<TaskWakeupMethWrapper object at 0x7efe3faafb88>()]>>
ERROR:asyncio:Task was destroyed but it is pending!
task: <Task pending coro=<Model._watch.<locals>._all_watcher() running at /home/thedac/dev/run/.tox/clients/lib/python3.6/site-packages/juju/model.py:808> wait_for=<Fut
ure pending cb=[<TaskWakeupMethWrapper object at 0x7efe3fa14af8>()]>>
ERROR:asyncio:Task was destroyed but it is pending!
task: <Task pending coro=<Event.wait() running at /usr/lib/python3.6/asyncio/locks.py:283> wait_for=<Future pending cb=[<TaskWakeupMethWrapper object at 0x7efe3face9a8>
()]> cb=[_wait.<locals>._on_completion() at /usr/lib/python3.6/asyncio/tasks.py:380]>
ERROR:asyncio:Task was destroyed but it is pending!
task: <Task pending coro=<Connection._pinger.<locals>._do_ping() done, defined at /home/thedac/dev/run/.tox/clients/lib/python3.6/site-packages/juju/client/connection.p
y:248> wait_for=<Future pending cb=[<TaskWakeupMethWrapper object at 0x7efe3faaf498>()]> cb=[_wait.<locals>._on_completion() at /usr/lib/python3.6/asyncio/tasks.py:380]
>
ERROR:asyncio:Task was destroyed but it is pending!
task: <Task pending coro=<Event.wait() done, defined at /usr/lib/python3.6/asyncio/locks.py:269> wait_for=<Future pending cb=[<TaskWakeupMethWrapper object at 0x7efe3fa
ce7c8>()]> cb=[_wait.<locals>._on_completion() at /usr/lib/python3.6/asyncio/tasks.py:380]>
ERROR:asyncio:Task was destroyed but it is pending!
task: <Task pending coro=<WebSocketCommonProtocol.recv() done, defined at /home/thedac/dev/run/.tox/clients/lib/python3.6/site-packages/websockets/protocol.py:299> wait
_for=<Future pending cb=[<TaskWakeupMethWrapper object at 0x7efe3ebb5c18>()]> cb=[_wait.<locals>._on_completion() at /usr/lib/python3.6/asyncio/tasks.py:380]>
Exception ignored in: <generator object Queue.get at 0x7efe3ecae990>
Traceback (most recent call last):
  File "/usr/lib/python3.6/asyncio/queues.py", line 169, in get
    getter.cancel()  # Just in case getter is not done yet.
  File "/usr/lib/python3.6/asyncio/base_events.py", line 575, in call_soon
    self._check_closed()
  File "/usr/lib/python3.6/asyncio/base_events.py", line 358, in _check_closed
    raise RuntimeError('Event loop is closed')
RuntimeError: Event loop is closed
ERROR:asyncio:Task was destroyed but it is pending!
task: <Task pending coro=<Queue.get() done, defined at /usr/lib/python3.6/asyncio/queues.py:155> wait_for=<Future cancelled> cb=[_wait.<locals>._on_completion() at /usr
/lib/python3.6/asyncio/tasks.py:380]>
Exception ignored in: <coroutine object AllWatcherFacade.Next at 0x7efe3ee6b6d0>
Traceback (most recent call last):
  File "/home/thedac/dev/run/.tox/clients/lib/python3.6/site-packages/juju/client/facade.py", line 412, in wrapper
    reply = await f(*args, **kwargs)
  File "/home/thedac/dev/run/.tox/clients/lib/python3.6/site-packages/juju/client/_client1.py", line 59, in Next
    reply = await self.rpc(msg)
  File "/home/thedac/dev/run/.tox/clients/lib/python3.6/site-packages/juju/client/overrides.py", line 105, in rpc
    result = await self.connection.rpc(msg, encoder=TypeEncoder)
  File "/home/thedac/dev/run/.tox/clients/lib/python3.6/site-packages/juju/client/connection.py", line 308, in rpc
    result = await self._recv(msg['request-id'])
  File "/home/thedac/dev/run/.tox/clients/lib/python3.6/site-packages/juju/client/connection.py", line 210, in _recv
    return await self.messages.get(request_id)
  File "/home/thedac/dev/run/.tox/clients/lib/python3.6/site-packages/juju/utils.py", line 61, in get
    value = await self._queues[id].get()
  File "/usr/lib/python3.6/asyncio/queues.py", line 169, in get
    getter.cancel()  # Just in case getter is not done yet.
  File "/usr/lib/python3.6/asyncio/base_events.py", line 575, in call_soon
    self._check_closed()
  File "/usr/lib/python3.6/asyncio/base_events.py", line 358, in _check_closed
    raise RuntimeError('Event loop is closed')
RuntimeError: Event loop is closed
ERROR:asyncio:Task was destroyed but it is pending!
task: <Task pending coro=<AllWatcherFacade.Next() done, defined at /home/thedac/dev/run/.tox/clients/lib/python3.6/site-packages/juju/client/facade.py:409> wait_for=<Fu
ture cancelled> cb=[_wait.<locals>._on_completion() at /usr/lib/python3.6/asyncio/tasks.py:380]>
ERROR:asyncio:Task was destroyed but it is pending!
task: <Task pending coro=<Event.wait() done, defined at /usr/lib/python3.6/asyncio/locks.py:269> wait_for=<Future pending cb=[<TaskWakeupMethWrapper object at 0x7efe3fb
639d8>()]> cb=[_wait.<locals>._on_completion() at /usr/lib/python3.6/asyncio/tasks.py:380]>

**Comment by fnordahl** [comment]
Fixed in #139

**Comment by AurelienLourot** [comment]
This just happened to me while running the functional tests on https://github.com/openstack/charm-keystone/commit/a85ae0e852657adf4a630511ce131dc467ada473

```
2019-12-16 17:14:01 [ERROR] Fatal write error on socket transport
protocol: <asyncio.sslproto.SSLProtocol object at 0x7fa8ca28df98>
transport: <_SelectorSocketTransport fd=10>
Traceback (most recent call last):
  File "/usr/lib/python3.6/asyncio/selector_events.py", line 752, in write
    n = self._sock.send(data)
OSError: [Errno 9] Bad file descriptor
2019-12-16 17:14:01 [ERROR] Fatal error on SSL transport
protocol: <asyncio.sslproto.SSLProtocol object at 0x7fa8ca28df98>
transport: <_SelectorSocketTransport closing fd=10>
Traceback (most recent call last):
  File "/usr/lib/python3.6/asyncio/selector_events.py", line 752, in write
    n = self._sock.send(data)
OSError: [Errno 9] Bad file descriptor

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/lib/python3.6/asyncio/sslproto.py", line 650, in _process_write_backlog
    self._transport.write(chunk)
  File "/usr/lib/python3.6/asyncio/selector_events.py", line 756, in write
    self._fatal_error(exc, 'Fatal write error on socket transport')
  File "/usr/lib/python3.6/asyncio/selector_events.py", line 634, in _fatal_error
    self._force_close(exc)
  File "/usr/lib/python3.6/asyncio/selector_events.py", line 646, in _force_close
    self._loop.call_soon(self._call_connection_lost, exc)
  File "/usr/lib/python3.6/asyncio/base_events.py", line 591, in call_soon
    self._check_closed()
  File "/usr/lib/python3.6/asyncio/base_events.py", line 377, in _check_closed
    raise RuntimeError('Event loop is closed')
RuntimeError: Event loop is closed
2019-12-16 17:14:01 [ERROR] Task was destroyed but it is pending!

...

___________________________________ summary ____________________________________
  func: commands succeeded
  congratulations :)
```

[charm-keystone fetches the latest zaza](https://github.com/openstack/charm-keystone/blob/a85ae0e852657adf4a630511ce131dc467ada473/test-requirements.txt#L17) which should contain your fix. @fnordahl @thedac Could it be that the issue is back? Should we re-open it?
